Restore removed API from Extensions

ericstj opened the pull request:
Fixes https://github.com/dotnet/runtime/issues/87480
I spent some time seeing what it would look like to bring back API that was present on extensions packages from 2.1-3.1 so that libraries built against these old packages would at least bind against the latest packages.

I stubbed out most of the API that was added back, unless it seemed "easy" to make work or important.  

Have a look below and comment about the cases and if you think we should do more (or less).

This covers most of the cases.  There's one case I cannot fix:
```
CP0006: Cannot add interface member 'Microsoft.Extensions.Hosting.IHostBuilder.UseServiceProviderFactory<TContainerBuilder>(System.Func<Microsoft.Extensions.Hosting.HostBuilderContext, Microsoft.Extensions.DependencyInjection.IServiceProviderFactory<TContainerBuilder>>)' to e7\bin\Debug\net4.8\Microsoft.Extensions.Hosting.Abstractions.dll because it does not exist on e2\bin\Debug\net4.8\Microsoft.Extensions.Hosting.Abstractions.dll
```
This is an added method to an interface.  One way to _fix_ this would be to go back and modify any type that implemented that interface to add the method as "public virtual" (luckily the signature doesn't require any new types).  I was able to test this and it is sufficient for the runtime to accept that method as satisfying the interface -- even if the implementer didn't see that method as part of the interface at compile time.  Chances are [there are very few implementations](https://grep.app/search?q=class%20.%2AIHostBuilder&regexp=true) of `IHostBuilder` out there - so it might be tractable to patch them (or even just ignore them).

ericstj committed:
Restore missing API from Extensions

joperezr commented:
FYI @geeknoid

joperezr commented on src/libraries/Microsoft.Extensions.Caching.Memory/ref/Microsoft.Extensions.Caching.Memory.cs:
```diff
@@ -42,6 +42,8 @@ public partial class MemoryCacheOptions : Microsoft.Extensions.Options.IOptions<
         public MemoryCacheOptions() { }
         public Microsoft.Extensions.Internal.ISystemClock? Clock { get { throw null; } set { } }
         public double CompactionPercentage { get { throw null; } set { } }
+        [System.Obsolete("This is obsolete and will be removed in a future version.")]
```
Is this just the default obsolete message or do we actually mean the fact that we want to remove in the future?

ericstj replied:
This was the message that was on these members before they were removed.

joperezr commented on src/libraries/Microsoft.Extensions.Hosting/ref/Microsoft.Extensions.Hosting.cs:
```diff
@@ -128,6 +128,7 @@ public partial class ApplicationLifetime : Microsoft.Extensions.Hosting.IApplica
     [System.Runtime.Versioning.UnsupportedOSPlatformAttribute("tvos")]
     public partial class ConsoleLifetime : Microsoft.Extensions.Hosting.IHostLifetime, System.IDisposable
     {
+        public ConsoleLifetime(Microsoft.Extensions.Options.IOptions<Microsoft.Extensions.Hosting.ConsoleLifetimeOptions> options, Microsoft.Extensions.Hosting.IHostEnvironment environment, Microsoft.Extensions.Hosting.IHostApplicationLifetime applicationLifetime) { }
```
NIT: Should this have an obsolete suggesting to use the other two ctors?

joperezr commented on src/libraries/Microsoft.Extensions.Logging.Console/ref/Microsoft.Extensions.Logging.Console.cs:
```diff
@@ -9,14 +9,30 @@ namespace Microsoft.Extensions.Logging
         [System.Diagnostics.CodeAnalysis.RequiresDynamicCodeAttribute("Binding TOptions to configuration values may require generating dynamic code at runtime.")]
         [System.Diagnostics.CodeAnalysis.RequiresUnreferencedCodeAttribute("TOptions's dependent types may have their members trimmed. Ensure all required members are preserved.")]
-        public static Microsoft.Extensions.Logging.ILoggingBuilder AddConsoleFormatter<[System.Diagnostics.CodeAnalysis.DynamicallyAccessedMembersAttribute(System.Diagnostics.CodeAnalysis.DynamicallyAccessedMemberTypes.PublicConstructors)] TFormatter, [System.Diagnostics.CodeAnalysis.DynamicallyAccessedMembers(System.Diagnostics.CodeAnalysis.DynamicallyAccessedMemberTypes.All)] TOptions>(this Mic ...
+        public static Microsoft.Extensions.Logging.ILoggingBuilder AddConsoleFormatter<TFormatter, TOptions>(this Microsoft.Extensions.Logging.ILoggingBuilder builder) where TFormatter : Microsoft.Extensions.Logging.Console.ConsoleFormatter where TOptions : Microsoft.Extensions.Logging.Console.ConsoleFormatterOptions { throw null; }
```
NIT: I'm assuming the removal of the trimmer attributes is unintentional?

tarekgh commented:
Is it possible to have all re-added obsolete APIs be in one source file inside each component instead of scattering it across the code? using partial classes can do it I guess.

ericstj commented:
> Is it possible to have all re-added obsolete APIs be in one source file inside each component instead of scattering it across the code? using partial classes can do it I guess.

I did demonstrate this method in some of the assemblies, but I'm not sold on that being preferable.  Also if you noticed - most of the stuff we brought back was entire new types, very few members removed or renamed.

Rather than look at the content / actual mechanism I obsoleted code, have a look at the entire mess of types we're bringing back and weigh in on which we think would actually be meaningful to bring back.

IMO all the logging stuff looks pretty busted.  There is a lot we had to bring back and none of it "works" -- except for the internal types I had to make public.  I have a hard time believing that folks which depended on this would be unblocked by adding back throwing API.  Perhaps I could 1) invest more time in understanding if we could make it functional, 2) make it noop instead of throw, 3) cut bringing back all the logging.

Everything but logging looks pretty "easy".  We can bring it back in a way that seems likely to unblock folks that used it.

ericstj commented on src/libraries/Microsoft.Extensions.Logging.Console/src/ConsoleLogger.cs:
```diff
@@ -13,7 +14,8 @@ namespace Microsoft.Extensions.Logging.Console
     [UnsupportedOSPlatform("browser")]
-    internal sealed class ConsoleLogger : ILogger
+    [EditorBrowsable(EditorBrowsableState.Never)]
+    public sealed class ConsoleLogger : ILogger
```
These types have a ton of breaking changes, so it might be better to just expose empty types and rename these internal ones, or not bring back any of logging.

ericstj replied:
I wonder if the balance we might strike in logging is not to expose any of the concrete loggers or related helper types/interfaces.  Only expose the extension methods.  I might prototype this to see if that's doable.  Still looking for usage information to validate this approach.

ericstj replied:
This is the approach I'm taking.  Only bringing back entry-point APIs and not these pub-ternal like types.

ericstj committed:
Fix a couple missing interfaces

geeknoid commented on src/libraries/Microsoft.Extensions.Caching.Memory/src/MemoryCacheOptions.cs:
```diff
@@ -42,6 +42,9 @@ public class MemoryCacheOptions : IOptions<MemoryCacheOptions>
             }
         }
 
+        [System.Obsolete("This is obsolete and will be removed in a future version.")]
```
In general, is this the right message? Maybe instead:

"Present for backward compatibility only, please do not use in new code"

?

ericstj replied:
None of these messages have been scrubbed - I just brought back them as they were in the last release or put a TODO in place.  This PR isn't meant to review for style/final merge.  It's sharing an outline to drive discussion on next steps.

geeknoid commented on src/libraries/Microsoft.Extensions.Configuration/src/ConfigurationSection.cs:
```diff
@@ -30,6 +30,14 @@ public ConfigurationSection(IConfigurationRoot root, string path)
             _path = path;
         }
 
+        /// <summary>
+        /// Initializes a new instance.
+        /// </summary>
+        /// <param name="root">The configuration root.</param>
+        /// <param name="path">The path to this section.</param>
+        public ConfigurationSection(ConfigurationRoot root, string path) : this((IConfigurationRoot)root, path)
```
```suggestion
        [Obsolete]
        public ConfigurationSection(ConfigurationRoot root, string path) : this((IConfigurationRoot)root, path)
```

ericstj replied:
We cannot add obsolete here since it would be source-breaking.  The compiler would prefer the concrete type over interface overload.  I don't think there is any harm in having this overload.

https://sharplab.io/#v2:EYLgxg9gTgpgtADwGwBYA0AXEBLANmgExAGoAfAAQCYBGAWACgGA3AQygAIx2BedgOxgB3dgGEAFAEoA3A2Zt2wHvyHsAQmLDTZjeuQDM7KqPYh2ASQBiLMDAYBvBgF9t+9tj4YYUAGbWY5qxt2O2cdVyNVewZ2GMMDdUs/dl8bCWDQ2PYAbQB5YABnCFwYTwBdaNjXdRFONNMMAAtsfLExRNTNdO1HIA===

geeknoid commented on src/libraries/Microsoft.Extensions.Hosting.Abstractions/src/IApplicationLifetime.cs:
```diff
@@ -14,30 +14,6 @@ namespace Microsoft.Extensions.Hosting
-        void StopApplication();
-    }
+    public interface IApplicationLifetime : IHostApplicationLifetime
+    { }
```
Isn't deleting all these members a breaking change?

ericstj replied:
They aren't deleted, they are "pushed down".  What happened in the past was the interface was renamed.  At first I added them back but then realized I couldn't use them in other API.  So I instead added them back as derived from their new name.  We use this trick in classes - pushing a member to a base class is a compatible change (both at compile and runtime).  I tested and this works for implicitly implemented interfaces, but it looks like the runtime isn't happy with it if the interface is explicitly implemented.  Probably that could be fixed, but not in the places we care about.  That makes this approach a no-go.

I could do the same thing with adapters.  Alternatively we could just throw from API that used these interfaces and not try to implement it.

ericstj committed:
Fix EventLog test

ericstj committed:
Undo interface heirarchy change in hosting

We can't push members down in interfaces.  While this works for implicit
interface implementations, the runtime will fail the type load if the
type explicitly implemented those members in the old interface.

ericstj committed:
Add API compat validation for old Extensions baseline

ericstj committed:
Bring back less logging API and baseline missing

ericstj committed:
Implement obsolete console API

ericstj committed:
Fix the download of extensions compat packages

geeknoid commented:
@RussKie has been working to get R9 samples working and encountering compat issues related to the fact old R9 had graduated dependencies (where the R9 bits for net462 depended on MS extensions 2.2, but the R9 bits for net7.0 depended on MS extensions 7.0, etc.), whereas the newer R9 now depends exclusively on MS extensions 8.0.

So these R9 samples could be used as a good basic vetting of whether the compat fixes actually make stuff work....

ericstj commented:
> So these R9 samples could be used as a good basic vetting of whether the compat fixes actually make stuff work....

I looked at this failure and its ASP.NETCore 2.x on .NET Framework, I suspect all the blockers will originate from this scenario and less R9's usage of extensions.  I was able to unblock it with the bits from this PR, but not certain about the amount of coverage that sample has.  Also not sure if we want to focus on this scenario for .NET consumption - it might be better to show usage in classic ASP.NET since that's where more customers will be on that framework.

ericstj commented on src/libraries/Microsoft.Extensions.Hosting/src/Internal/ConsoleLifetime.obsolete.cs:
```diff
@@ -0,0 +1,66 @@
+    public partial class ConsoleLifetime : IHostLifetime
+    {
+        [Obsolete("IHostingEnvironment and IApplicationLifetime have been deprecated. Use Microsoft.Extensions.Hosting.IHostEnvironment and IHostApplicationLifetime instead.")]
+        public ConsoleLifetime(IOptions<ConsoleLifetimeOptions> options, IHostingEnvironment environment, IApplicationLifetime applicationLifetime)
```
This addition is causing many tests to fail with: 
```
Void .ctor(Microsoft.Extensions.Options.IOptions`1[Microsoft.Extensions.Hosting.ConsoleLifetimeOptions], Microsoft.Extensions.Hosting.IHostEnvironment, Microsoft.Extensions.Hosting.IHostApplicationLifetime, Microsoft.Extensions.Options.IOptions`1[Microsoft.Extensions.Hosting.HostOptions], Microsoft.Extensions.Logging.ILoggerFactory)
Void .ctor(Microsoft.Extensions.Options.IOptions`1[Microsoft.Extensions.Hosting.ConsoleLifetimeOptions], Microsoft.Extensions.Hosting.IHostingEnvironment, Microsoft.Extensions.Hosting.IApplicationLifetime)
```

ericstj replied:
Given this type is internal, and API usage is literally non-existent I'm going to pull this one out:
https://apisof.net/catalog/74559930-0e33-2f5b-82ad-66e031022b84

ericstj committed:
Remove ConsoleLifetime overload

This method was creating a binding ambiguity for DI.
It has zero usage, so just exclude it.

ericstj committed:
Update obsolete message on InplaceStringBuilder

ericstj committed:
Fill in implementation of obsolete logging methods

ericstj committed:
Update console->config ref dependency

ericstj commented:
Some information about using these packages on different versions of .NET.

| Framework | In support? | Works with 8.0 packages? | Works with this PR? | Notes |
| ------------ | ------------ | ----------------------------- | ---------------------- | ------ |
| NETFramework (net462+) | yes | yes | yes | No dependencies on extensions by default | 
| ASP.NETCore on NETFramework (net462+) | yes* | no | yes | While this does enable upgrade of extensions packages here, we  document that we don't support that upgrade https://dotnet.microsoft.com/en-us/platform/support/policy/aspnet/2.1-packages | 
| netcoreapp2.1 | no | no | yes | Avoids problem with 3.x due to not calling WebHost instead of generic host APIs |
| netcoreapp3.1 | no | no | somewhat | These packages allow the app to get further, but it fails to start on Windows due to EventLog dropping support for netcoreapp3.1 | 
| net5.0 | no | somewhat | somewhat | Fails to start on Windows for same reason as 3.1 | 
| net6.0 | yes | yes | yes | |
| net7.0 | yes | yes | yes | |

So really the only place that this PR is making a difference is on unsupported frameworks or for unsupported upgrades.

geeknoid commented:
But the point of these changes is not to show that net462 code can work with .NET 8 packages, it's about making it possible for an application developed using 2.2 extensions will work when the 8.0 extension are substituted under. If there are breaking changes between 2.2 and 8.0, then this is not possible.

So can a net462 app compiled using the 2.2 version of MS Extensions work with the 8.0 version of MS Extensions?

ericstj commented:
> So can a net462 app compiled using the 2.2 version of MS Extensions work with the 8.0 version of MS Extensions?

It depends on what they used.  If they didn't use any of the obsoleted API it'd be fine.  We need to see what "typical" usage of extensions looks like on net462 that is not ASP.NET Core on NETFramework (which is the only reason for folks to stay on 2.1 -- in which case we are not interested in updating them per above).  Also - 2.2 packages have been out of support for many years now.  There's really no reason for folks to stay on those versions.  There is no "platform affinity" between NETFramework versions and any specific release of Microsoft.Extensions nuget packages, so we shouldn't be thinking about pulling folks forward on NETFramework - those are just the latest packages with the latest features and fixes - just like any other package.

ericstj committed:
Update Obsolete messages and make error

ericstj committed:
Suppress compat around Obsolete(..., error: true)

ericstj committed:
Fix more obsolete messages

ericstj committed:
Remove ConsoleLoggerProvider ctor overloads

These don't have much usage and their addition could create problems for
DI Providers.

ericstj committed:
Omit ConfigurationSection .ctor overload

This overload has a source compatible replacement, and is only used by
one package in all of nuget.org - which already has assets targeting new
TargetFrameworks which don't depend on this API.

ericstj committed:
Remove OptionsWrapper members

These were only available as non-obsolete in 2.0 and there is zero usage
on nuget.org.

ericstj committed:
Mark restored obsoleted API as EditorBrowsable=never

ericstj committed:
Remove extensions 2.1 compat infrastructure

ericstj commented:
This is now ready for review.  The final decision on in/out API and justification can be found here:
[MSFT internal xls](https://microsoft.sharepoint.com/:x:/r/teams/netfx/corefx/_layouts/15/Doc.aspx?sourcedoc=%7BC95246F3-E8C6-4911-9B27-B99C9BEEEE85%7D&file=extensions-compat-usage.xlsx&action=default&mobileredirect=true&share=IQHzRlLJxugRSZsnuZyb7u6FAQe-PcNae_gSYRixThXzBmY) / [public gist](https://gist.github.com/ericstj/c06988661190ec62736c748316692675)
You can also see the differences with what we didn't bring back in [this commit](https://github.com/dotnet/runtime/pull/85846/commits/5bdf838de2ccfffae79cbdd752b37030ab5ef2a0) from before I removed that compat infra.

joperezr commented:
> This is now ready for review. The final decision on in/out API and justification can be found here: [MSFT internal xls](https://microsoft.sharepoint.com/:x:/r/teams/netfx/corefx/_layouts/15/Doc.aspx?sourcedoc=%7BC95246F3-E8C6-4911-9B27-B99C9BEEEE85%7D&file=extensions-compat-usage.xlsx&action=default&mobileredirect=true&share=IQHzRlLJxugRSZsnuZyb7u6FAQe-PcNae_gSYRixThXzBmY) / [public gist](https://gist.github.com/ericstj/c06988661190ec62736c748316692675) You can also see the differences with what we didn't bring back in [this commit](https://github.com/dotnet/runtime/pull/85846/commits/5bdf838de2ccfffae79cbdd752b37030ab5ef2a0) from before I removed that compat infra.

This is awesome, thanks a bunch for driving this @ericstj.

@geeknoid can we do a quick skim on that xls/gist to make sure all of the decisions there make sense?

@RussKie I know that you were previously doing a porting exercise of a sample that required some of these APIs to be brought back, and you used a private build at the time that Eric shared with you. Can we try that exercise again with the latest version of the packages from this PR to ensure everything still works as expected?

ericstj commented:
You can find a private build of this PR from https://devdiv.visualstudio.com/DevDiv/_artifacts/feed/ericstj-share with the version `8.0.0-preview.6.extCompat.1`

ericstj commented:
@tarekgh can you have a look at the final shape the Source changes are in and let me know if you’d like anything to change?

tarekgh commented on src/libraries/Microsoft.Extensions.Logging.Console/src/ConsoleLoggerExtensions.Obsolete.cs:
```diff
@@ -0,0 +1,126 @@
+            }
+        }
+
+        private sealed class NullChangeToken : IChangeToken, IDisposable
```
>NullChangeToken [](http://example.com/codeflow?start=29&length=15)

do we need to mark this as obsolete too? I am seeign we did that with `ConsoleLoggerSettingsAdapter`

ericstj replied:
No, it's private.  I marked `ConsoleLoggerSettingsAdapter` obsolete because it needed to use other obsolete API.  This type does not, so it doesn't need obsolete.

tarekgh commented on src/libraries/Microsoft.Extensions.Logging.Debug/src/DebugLogger.cs:
```diff
@@ -31,8 +31,7 @@ public DebugLogger(string name)
         {
-            // If the filter is null, everything is enabled
-            // unless the debugger is not attached
+            // Everything is enabled unless the debugger is not attached
```
>Everything [](http://example.com/codeflow?start=15&length=10)

nit: I agree the old comment is unclear but it was providing extra info that LogLevel.Debug is the highest level before `None` and that what refer to filter in the old comment.

ericstj replied:
I don't think so.  This type used to have a member named filter.  That was removed but the comment was not updated: https://github.com/dotnet/extensions/blob/ab493da045ef1e645ff52675adb177ba7a503ce0/src/Logging/Logging.Debug/src/DebugLogger.cs#L46-L50

tarekgh commented on src/libraries/Microsoft.Extensions.Options/src/OptionsWrapper.cs:
```diff
@@ -1,6 +1,8 @@
 
+using System;
+using System.ComponentModel;
 using System.Diagnostics.CodeAnalysis;
```
why adding these here? wasn't the code compiling before adding them?

ericstj replied:
These were here from when I previously added obsolete API.  Missed cleaning these up.  Thanks!

tarekgh commented on src/libraries/Microsoft.Extensions.Primitives/src/InplaceStringBuilder.cs:
```diff
@@ -0,0 +1,135 @@
+
+namespace Microsoft.Extensions.Primitives
+{
+    [DebuggerDisplay("Value = {_value}")]
```
>[DebuggerDisplay("Value = {_value}")] [](http://example.com/codeflow?start=4&length=37)

it is interesting to keep the debug display while we are obsoleting the struct. I am fine either way though.

ericstj replied:
I just pulled this file exactly as it was from the last shipping release.  https://github.com/dotnet/extensions/blob/release/3.1/src/Primitives/src/InplaceStringBuilder.cs

tarekgh commented on src/libraries/Microsoft.Extensions.Primitives/src/InplaceStringBuilder.cs:
```diff
@@ -0,0 +1,135 @@
+                }
+
+                // _offset > 0 indicates writing state
+                if (_offset > 0)
```
>_offset [](http://example.com/codeflow?start=20&length=7)

this is interesting as it is not thread safe :-)

ericstj replied:
This type is not thread safe.

tarekgh reviewed (approved):
LGTM. nice to see returning the compatibility back. 

Is there anything else left we decided not to bring?

ericstj commented:
> Is there anything else left we decided not to bring?

The XLS and gist linked above show what we didn't bring back.

ericstj committed:
Remove unused usings
